wagtail/core/query.py

```diff
@@ -179,43 +198,61 @@ def not_page(self, other):
         """
         return self.exclude(self.page_q(other))
 
-    def type_q(self, klass):
-        content_types = ContentType.objects.get_for_models(*[
+    def type_q(self, klass, excluding=False):
+        models = [
             model for model in apps.get_models()
             if issubclass(model, klass)
-        ]).values()
-
-        return Q(content_type__in=list(content_types))
+        ]
+        if not excluding:
+            self._specified_classes = models
+        content_types = ContentType.objects.get_for_models(*models)
+        return Q(content_type__in=list(content_types.values()))
 
     def type(self, model):
         """
         This filters the QuerySet to only contain pages that are an instance
         of the specified model (including subclasses).
         """
-        return self.filter(self.type_q(model))
+        return self.filter(self.type_q(model, excluding=False))
 
     def not_type(self, model):
         """
         This filters the QuerySet to not contain any pages which are an instance of the specified model.
         """
-        return self.exclude(self.type_q(model))
+        return self.exclude(self.type_q(model, excluding=True))
 
-    def exact_type_q(self, klass):
+    def exact_type_q(self, klass, excluding=False):
+        if not excluding:
+            self._specified_classes = [klass]
         return Q(content_type=ContentType.objects.get_for_model(klass))
 
     def exact_type(self, model):
         """
         This filters the QuerySet to only contain pages that are an instance of the specified model
         (matching the model exactly, not subclasses).
         """
-        return self.filter(self.exact_type_q(model))
+        return self.filter(self.exact_type_q(model, excluding=False))
 
     def not_exact_type(self, model):
         """
         This filters the QuerySet to not contain any pages which are an instance of the specified model
         (matching the model exactly, not subclasses).
         """
-        return self.exclude(self.exact_type_q(model))
+        return self.exclude(self.exact_type_q(model, excluding=True))
+
+    def child_of(self, parent_page):
+        """
+        This filters the QuerySet to only contain pages that are direct
+        children of the specified page.
+
+        Data from the parent page's class is used to limit the 'upcast'
+        lookups, to only include joins for page types that are allowed
+        to exist in this part of the tree.
+        """
+        parent_class = parent_page.specific_class
+        if parent_class is not None:
+            self._specified_classes = parent_class.allowed_subpage_models()
```

Comment:
I don't think we can rely on `allowed_subpage_models` for covering all of the possible models. This is only ever checked by the Wagtail admin and they can be changed over time without any cleanup.

Reply:
That is a shame :( child_of() is probably the most commonly used filter (especially within wagtail itself) - and it provides an excellent opportunity simplify the query. when `subpage_classes` is used well.

Reply:
I suppose we could do a little query to identify all content_types used at this level, but I would guess that in sites with quite a small number of page types, that might be less efficient than just including joins for all types.

Reply:
If you don't mind a bit of complexity, we could possibly use a bloom filter for this. 

This would mean adding a 32-bit integer on each page. Whenever a new page is created, we set a bit on the parent page's bloom filter corresponding to the ID of the content type we just created wrapped to 32 bits. (eg `parent.bloom |= 1 << (new_page.content_type_id % 32)`)

Then in the query, you do a little calculation to find which page types have content type IDs that would match that filter.

I know it's a bit complicated to what we would usually do, but could give up to a 32x improvement if these `JOIN`s turn out to slow things down.

Reply:
@kaedroho I do really like this idea - However, I think I'd quite like to tackle that in a separate PR, as this one is already likely to get a little 'sprawly' once I start digging in to remove unnecessary uses of `specific()`. Sounds like there would be some additional considerations there too, like what to do for existing pages, so probably best tackled separately.